src/lib.rs

```diff
@@ -157,145 +178,157 @@ pub trait ASTRule: Rule {
-    line: usize,
-    col: usize,
+    range: &TextRange,
     message: &str,
     path: &ColoredString,
     code: &ColoredString,
 ) -> fmt::Result {
-    let lines = read_lines(&diagnostic.path);
-    let mut start_index = line.saturating_sub(2).max(1);
+    let source_code = diagnostic.file.to_source_code();
+    let content_start_index = source_code.line_index(range.start());
+    let mut start_index = content_start_index.saturating_sub(2);
 
     // Trim leading empty lines.
-    while start_index < line {
-        if !lines[start_index.saturating_sub(1)].trim().is_empty() {
+    while start_index < content_start_index {
+        if !source_code.line_text(start_index).trim().is_empty() {
             break;
         }
         start_index = start_index.saturating_add(1);
     }
 
-    let mut end_index = line.saturating_add(2).min(lines.len());
+    let content_end_index = source_code.line_index(range.end());
+    let mut end_index = content_end_index
+        .saturating_add(2)
+        .min(OneIndexed::from_zero_indexed(source_code.line_count()));
 
     // Trim leading empty lines.
-    while end_index > line {
-        if !lines[end_index.saturating_sub(1)].trim().is_empty() {
+    while end_index > content_end_index {
+        if !source_code.line_text(end_index).trim().is_empty() {
             break;
         }
         end_index = end_index.saturating_sub(1);
     }
 
-    let content_slice = lines[start_index.saturating_sub(1)..end_index]
-        .iter()
-        .fold(String::default(), |acc, line| format!("{acc}{line}\n"));
+    let start_offset = source_code.line_start(start_index);
+    let end_offset = source_code.line_end(end_index);
 
-    // Annotations are done by offset, so we need to count line
-    // lengths... including the newline character, which doesn't
-    // appear in `lines`!
-    let offset_up_to_line = lines[start_index.saturating_sub(1)..line.saturating_sub(1)]
-        .iter()
-        .fold(0, |acc, line| acc + line.chars().count() + 1);
+    let source = source_code.slice(TextRange::new(start_offset, end_offset));
+    let message_range = range - start_offset;
 
-    // Something really weird going on here, where I can't get it to
-    // put the annotation in the first column: it's either in column 2
-    // or the end of the previous line. But does appear to be right
-    // for other columns!
-    let label_offset = offset_up_to_line + col.saturating_sub(1);
+    let start_char = source[TextRange::up_to(message_range.start())]
+        .chars()
+        .count();
+    let end_char = source[TextRange::up_to(message_range.end())]
+        .chars()
+        .count();
 
     // Some annoyance here: we *have* to have some level prefix to our
     // message. Might be fixed in future version of annotate-snippets
     // -- or we use an earlier version with more control.
     // Also, we could use `.origin(path)` to get the filename and
     // line:col automatically, but see above about off-by-one error
```

Comment:
This refers back to earlier comments that were removed. Can this go too, or does it need to be reworded?